ENH: Added "kappa4" and "kappa3" distributions.

timcera opened the pull request:
Added "kappa4" and "kappa3" distributions to
scipy/stats/_continuous_distns.py.  Added some additional tests for
"kappa4" for parameters that should equal other distributions.

josef-pkt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,114 @@ def _ppf(self, q, k, s):
+                      (1.0 - k*x)**(1.0/k - 1.0)*np.exp(-(1.0
+                                                          - k*x)**(1.0/k)),
+                      np.exp(-x)*(1.0 - h*np.exp(-x))**(1.0/h - 1),
+                      np.exp(-x)*np.exp(-np.exp(-x))]
```
looks like this would be better implemented via _logpdf

josef-pkt replied:
also, without lazy_where this might be relatively expensive because it calculates all versions.

timcera replied:
I realize that it is inelegant.  I just mapped equations and concepts from literature.  

I would need some guidance on translation of these equations and h and k to _logpdf.  I did a similar thing for the pearson3 distribution, so it is not impossible.

timcera replied:
Didn't know about _lazywhere.  Easy change.

josef-pkt replied:
does it work for this case? I don't remember. 
Or, does it need to be extended to handle more than two segments/alternatives?

ev-br replied:
IMO it'd be cleanest to add _lazyselect

timcera replied:
_lazyselect sounds like a great idea, but I couldn't find it.  Is there anything already available that I missed?

I played around with a couple lazyish implementations but they ended up being pretty ugly to look at.

timcera replied:
Just added some more tests and realize that I have to have something that has lazy evaluation.  Will try to work out an implementation either using _lazywhere or rolling my own.

ev-br replied:
I remember trying to use _lazywhere for three-day dispatch and it was already a mess. So better add one for a list of conditions and callables and hide the uglyness in it. And it'll be immediately useful in e.g. trapz.

josef-pkt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,114 @@ def _ppf(self, q, k, s):
+
+    """
+    def _argcheck(self, h, k):
+        return (h >= -2) & (h <= 2) & (k >= -2) & (k <= 2)
```
does this need restriction for bounded support, lower or upper limit ?  `.a` `.b`
should be parameter dependent,AFAICS

timcera replied:
Indeed, there are upper and lower bounds dependent on h and k, but at least as far as I could figure out it would be difficult to implement in scipy since loc and scale are part of the transformed x before any of the "_*" methods are called.

ξ + α ( 1 – h –k) / k ≤ x ≤ ξ + α/k    if h > 0 & k > 0
ξ + α ln(h) ≤ x ≤ ∞                          if h > 0 & k = 0
ξ + α ( 1 – h –k) / k ≤ x ≤ ∞            if h > 0 & k < 0
– ∞ < x ≤ ξ + α/k                            if h ≤ 0 & k > 0
– ∞ < x < ∞                                    if h ≤ 0 & k = 0
ξ + α/k ≤ x < ∞                               if h ≤ 0 & k < 0

from: http://etd.lsu.edu/docs/available/etd-05182004-144851/unrestricted/Finney_dis.pdf

josef-pkt replied:
You can look at genextreme (IIRC) or similar that have support depending on a shape parameter.
`loc` and `scale` adjustments are handled separately in a generic way, and AFAIR you don't have to worry about those in the individual distributions.

timcera replied:
Seems like "c" in genextreme defines limits on a "loc" and "scale" transformed "x".  This is not the case for the limit equations that I found.  For example: say h<=0 and k > 0, then the untransformed "x" will have limits – ∞ < x ≤ ξ.  Within _argcheck however I don't have access to ξ.  I could assume that ξ = 0, which is probably of most of the time, but then the limits would be incorrect if someone used a different ξ.

josef-pkt replied:
you need to define the limits for the transformed, standardized loc=0 and scale=1. The limits will be transformed to the original x with whatever loc and scale the user specifies.
(That's standard behavior for loc/scale transformation, and is not always the most obvious parameterization for individual distribution.)  I haven't looked at the specifics yet, but given that you have genextreme as a special case here, you will need something like the `c` for the support of the standardized distribution.

josef-pkt commented:
Interesting, I've never seen this one, but it looks like a useful generalization of several distributions.

timcera commented:
Trying to implement distributions used in my line of work.  Got pearson3 implemented in 2012.  That was a long time ago!  Kappa4 and Kappa3 is next, then Wakeby.  

Is there a list of distributions left to implement?

josef-pkt commented:
> Is there a list of distributions left to implement?

I looked at a matlab package once that had a huge number of distributions (mostly only pdf) implemented. 
I think only adding famous distributions and those distributions where someone has a specific usecase is the right approach for adding distributions. (Adding a few hundred "exotic" distributions might not be worth the effort.)

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,114 @@ def _ppf(self, q, k, s):
+
+        kappa3.pdf(x, a) = a*[a + x**a]**(-(a + 1)/a)
+
+    for ``a > 0``.
```
and x>=0

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,114 @@ def _ppf(self, q, k, s):
+        return a*(a + x**a)**(-(a + 1.0)/a)
+
+    def _cdf(self, x, a):
+        return x*(a + x**a)**(-1.0/a)
```
Can invert this to get the explicit form of the ppf.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,114 @@ def _ppf(self, q, k, s):
+
+    %(example)s
+
+    """
```
Needs
```
def _argcheck(self, a):
    retutn a>0
```

timcera replied:
This comment is for kappa3, and it was implemented a while ago, but github didn't keep it in the right place.

Fixed.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,114 @@ def _ppf(self, q, k, s):
+                    np.logical_and(h == 0, k == 0)]
+        choicelist = [1.0/k*(1.0 - ((1.0 - x**h)/h)**k),
+                      np.ones_like(x)/k,
+                      x*inf,
```
0*inf = nan

ev-br replied:
And how come it's not -log (-log x) if cdf=exp(-exp(-x))

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,142 @@ def _ppf(self, q, k, s):
+        hc = np.atleast_1d(h)[0]
+        self.hc = hc
+        kc = np.atleast_1d(k)[0]
+        self.kc = kc
```
I've to admit I don't like this. Granted, there does not seem to be a way of correctly handling e.g. `h=[0, 1]` --- the generic code path does not handle well the case where `a` or `b` are shape-dependent. 

Nonetheless, attaching attributes to the instance like this feels like a hack. More importantly, if and when there is work to allow handling array-valued support bounds, this hack will need to be undone *in addition to* fixing the generic code paths.

At the very least, these extra attributes must have a leading underscore so that they are not public.

All in all, my gut feeling is that it's better to use a helper along the lines of https://github.com/scipy/scipy/commit/0ffb413cf8184ba20d7c33cd7e1b1027d4c03473
(this `_lazyselect` is not very well tested, so can be buggy, too).

timcera replied:
Not sure what to do with self.a and self.b.  I currently have the support defined in a shape dependent way for both.

self.hc and self.kc are gone in my local version.

Now using _lazyselect for this and kappa4 _pdf, _cdf, and _ppf.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,142 @@ def _ppf(self, q, k, s):
+        elif hc != 0 and kc == 0:
+            return np.exp(-x)*(1.0 - h*np.exp(-x))**(1.0/h - 1.0)
+        elif hc == 0 and kc == 0:
+            return np.exp(-x)*np.exp(-np.exp(-x))
```
This is `gumberl_r._pdf(x)`, correct?

timcera replied:
Yes.  See table in docstring and included as one of the kappa4 tests.

ev-br commented on scipy/stats/_distr_params.py:
```diff
@@ -54,6 +54,11 @@
+    ['kappa4', (-0.1, 0.1)],
+    ['kappa4', (0.0, 0.1)],
+    ['kappa4', (-0.1, 0.0)],
+    ['kappa4', (0.0, 0.0)],
```
Minor: I *think* the html documentation will end up having an example with `h=k=0`. Maybe better reorder the entries.

timcera replied:
Reorder - no problem.

Is this the right way to test different domains of the distribution?  It looks like a good spot, but kappa4 is the only one so far to have multiple entries.

ev-br replied:
I think it's appropriate, yes. Something else has multiple entries, is it hypegeometric.

ev-br commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -679,6 +679,79 @@ def test_cdf(self):
+        # Couldn't get the generalized logistic test to work, I think
+        # mostly because of how to establish the "c" parameter for the
+        # general logistic distribution.
+        pass
```
Skimming http://etd.lsu.edu/docs/available/etd-05182004-144851/, (found from quick googling), it seems that `h=-1` and `k=0` matches genlogistic with `c=1`  (cf Eq. (13) in the reference above)

timcera replied:
Will try genlogistic with c=1.  Thanks.

ev-br commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -679,6 +679,79 @@ def test_cdf(self):
+        # h = 0.0 and k = 1.0 is reverse exponential
+
+        # scipy doesn't seem to have a reverse exponential distribution.
+        pass
```
What's so special about this "reverse exponential"? It seems to be just `cdf \propto exp(x)` for `-\infty < x < 1`.

timcera replied:
Finney mentions reverse exponential and I just wanted to include it as a test.  Nothing special beyond that.  I couldn't find an explicitly labeled distribution and I didn't look at anything beyond that.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,142 @@ def _ppf(self, q, k, s):
+        kappa4.pdf(x, h, k) = (1.0 - k*x)**(1.0/k - 1.0)*
+                              (1.0 - h*(1.0 - k*x)**(1.0/k))**(1.0/h-1)
+
+    if ``h`` and ``k`` are not equal to 0.
```
I think it also needs formulas for special cases, `h` or `k` being zero. To simplify formulas, you might also want to list cdfs not pdfs: eqs (1-4) in Finney look simpler than (11-14).

Domains of support should also be listed here.

timcera replied:
Already done and in my working version.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,142 @@ def _ppf(self, q, k, s):
+    +------+-------------+--------------+---------------+
+    | -1.0 | Logistic    |              | Generalized   |
+    |      |             |              | Logistic      |
+    +------+-------------+--------------+---------------+
```
Nice table! 
As noted elsewhere, I'm not sure about reverse exponential and generalized logistic though.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,142 @@ def _ppf(self, q, k, s):
 
 
+class kappa4_gen(rv_continuous):
+    """Kappa 4 parameter distribution
```
Nitpick: full stop at the end of line (otherwise, the doc builder gets confused IIRC)

timcera replied:
Done.

ev-br commented:
I've left a couple of comments inline. In addition, I think this needs:

* some references. The best I've found is a PhD thesis, http://etd.lsu.edu/docs/available/etd-05182004-144851/;

* some more testing, especially in and close to degenerate cases (cf e.g. genpareto).

* possibly some calculations are better in logspace --- again, testing in near-degenerate regimes should show this.

* Test failures seem real. I suspect there are issues with either formulas for pdf et al or with the support bounds.

ev-br commented:
Here's one of the test failures:

```
In [29]: h, k
Out[29]: (-0.1, 0.0)

In [30]: kappa4.expect(lambda x: 1, args=(h, k))
/home/br/repos/scipy/build/testenv/lib/python3.5/site-packages/scipy/integrate/quadpack.py:357: IntegrationWarning: The occurrence of roundoff error is detected, which prevents 
  the requested tolerance from being achieved.  The error may be 
  underestimated.
  warnings.warn(msg, IntegrationWarning)
Out[30]: nan
```

Move the computation to the log space:

```
In [31]: def kpdf1(x, h, k):
    ex = np.exp(-x)
    logval = (1.0/h - 1.0) * log1p(-h*ex)
    return np.exp(-x + logval)
   ....: 

In [32]: quad(lambda x: kpdf1(x, h, k), -np.inf, np.inf)
Out[32]: (1.0000000000000002, 5.700726692493948e-10)
```

ev-br commented:
Test failures for `kappa3` are also real. Here the issue is different though: an `m`-th moment of `kappa3`, as defined here, only exists for `m < a`: otherwise the integral of `x**m * pdf(x, a)` diverges at the upper limit. The test loop runs for `a=1`, and this kappa3 reduces to a Cauchy distribution, where moments do not exist.

A way to take care of this is to define `_stats` where you return `nan` if a moment does not exist or `None` if it does.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3247,6 +3247,142 @@ def _ppf(self, q, k, s):
+    if ``h`` and ``k`` are not equal to 0.
+
+    kappa4 takes ``h`` and ``k`` as shape parameters, skew and
+    kurtosis respectively.
```
```
In [37]: h = k = 0

In [38]: kappa4.stats(h, k, moments='mvsk')
Out[38]: 
(array(0.5772156649133174),
 array(1.6449340668314645),
 array(1.139547099046641),
 array(2.40000000155596))
```

timcera commented:
Evgeni, thank you for your review.  Fortunately I have most of your comments addressed in my working version, but am waiting until I pass all tests.  There is one test for kappa3 that is failing, and I couldn't include the fit tests for kappa4.

I am now using _lazyselect, but I didn't find your version, so I wrote my own - and it seems that both of us used _lazywhere as a model since they are nearly identical.

timcera commented:
I still have three 'check_normalization' tests that fail.  At this point I need some help.  I thought by now I would have tried _everything_ but that is obviously not true since I didn't try the thing that actually works.

ev-br commented:
Something fishy seems to be going on with the limits of support. 
To quickly check it, instrument the `expect` method of `rv_continuous` class with `import pdb; pdb.set_trace()`, then `$ python runtests.py --ipython`, and in the resulting ipython session do

```
In [5]: h, k = 0.1, 0

In [6]: kappa4.expect(lambda x: 1, args=(h, k))
> /home/br/repos/scipy/build/testenv/lib/python2.7/site-packages/scipy/stats/_distn_infrastructure.py(2313)expect()
-> olderr = np.seterr(all='ignore')
(Pdb) l
2308 	        kwds['args'] = args
2309 	
2310 	        import pdb; pdb.set_trace()
2311 	
2312 	        # Silence floating point warnings from integration.
2313 ->	        olderr = np.seterr(all='ignore')
2314 	        vals = integrate.quad(fun, lb, ub, **kwds)[0] / invfac
2315 	        np.seterr(**olderr)
2316 	        return vals
2317 	
2318 	
(Pdb) self.a, self.b
(-inf, inf)
(Pdb) integrate.quad(fun, np.log(0.1), np.inf, **kwds)
(0.9999999999999996, 9.488071859664648e-09)
```

This `a = -inf` does not look right: the has a factor of exp(-x) which blows up.

argriffing commented:
one of the failing tests is just a pep8 issue
```
scipy/stats/_continuous_distns.py:3418:35: E124 closing bracket does not match visual indentation
1       E124 closing bracket does not match visual indentation
```

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3261,6 +3262,313 @@ def _ppf(self, q, k, s):
+        def f5(h, k):
+            return 1.0/k
+
+        self._a = _lazyselect(condlist,
```
Here. Should be `self.a`, no underscore, here and elsewhere. And this takes care of the CI failures for me locally.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3261,6 +3262,313 @@ def _ppf(self, q, k, s):
+        return (h == h)
+
+    def _pdf(self, x, h, k):
+        cond = np.logical_and(x >= self._a, x <= self._b)
```
Once `self._a` is renamed to `self.a`, this condition is no longer needed I believe, because it's handled in the generic `pdf`.

ev-br commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -679,6 +679,85 @@ def test_cdf(self):
+
+        # The genlogistic currently in scipy doesn't match kappa4, h=-1,
+        # k != 0.
+        pass
```
Then this test should be removed IMO, and this point needs to be clarified in the docstring.

ev-br commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -679,6 +679,85 @@ def test_cdf(self):
+        # h = 0.0 and k = 1.0 is reverse exponential
+
+        # scipy doesn't seem to have a "reverse" exponential distribution.
+        pass
```
Likewise, this test does not seem to do anything useful.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3261,6 +3262,313 @@ def _ppf(self, q, k, s):
+        kappa4.pdf(x, h, k) = exp(-x)*exp(-exp(-x))
+
+    kappa4 takes ``h`` and ``k`` as shape parameters, skew and
+    kurtosis respectively.
```
This statement does not seem to be correct: at `h=k=0`, the distribution is not symmetric, hence at least skew != 0.

timcera replied:
That statement is from http://file.scirp.org/pdf/JWARP20121000009_14676002.pdf page 867.  Not saying that they are right, just that is where I got it from.

ev-br replied:
I'd say whether that statement implies a different definition of skew/kurtosis, or is it something those authors might want to write an erratum for, it's up to them. As far as I can see, this is not correct, and should not be included as written.

timcera replied:
No problem.  Deleted mention of skew and kurtosis.

ev-br commented:
Now the remaining ingredients CI failures are in the the docstring is somehow not a valid ReST. The doc build on CircleCI could be helpful, but it fails for some other reason --- maybe a rebase on master would help

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3282,6 +3283,306 @@ def _ppf(self, q, k, s):
+            maxr = 5
+
+        outputs = [None if r < maxr else nan for r in range(1, 5)]
+        return outputs[:]
```
This looks strange: `None` signals "moment exists, defer to a generic calculation". 
For `h=k=0`, the moments do exist, but this function returns nans. Not sure about other cases.

```
In [14]: stats.kappa4.stats(h, k, moments='m')
Out[14]: array(nan)

In [15]: stats.kappa4.expect(lambda x: x, args=(h, k))
Out[15]: 0.577215664901533

In [16]: h, k
Out[16]: (0, 0)
```

timcera replied:
Good catch, fixed and running tests.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3282,6 +3283,306 @@ def _ppf(self, q, k, s):
+        return _lazywhere(x > 0,
+                          (x, a),
+                          lambda x, a: a*(a + x**a)**(-1.0/a-1),
+                          0.0)
```
_lazywhere not needed here as long as `a=0` in the constructor.

timcera replied:
Confused.  Per _argcheck, a > 0

ev-br replied:
Ah, yes. Indeed. a-the-shape-parameter is a>0, as controlled by `_argcheck`. a-the-lower-bound of the support, i.e., `self.a`, is set to be zero when creating the instance. https://github.com/scipy/scipy/pull/6163/files/9061fe875ff1bd45aee52399e06b0e6743968212#diff-6e4287784ff3e7a5c6d89d442bea3fbbR3583
The latter makes sure that only `x>=0` is seen by `_pdf` and friends. (The corner case of `x=self.a` is a bit inconsistent ATM; there's https://github.com/scipy/scipy/pull/6246 to fix it)

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3282,6 +3283,308 @@ def _ppf(self, q, k, s):
+        The "fifth" one is the one kappa4 should match which isn't
+        implemented in scipy:
+        https://en.wikipedia.org/wiki/Talk:Generalized_logistic_distribution
+        http://www.mathwave.com/help/easyfit/html/analyses/distributions/gen_logistic.html
```
```
In [30]: x = np.linspace(-10, 10, 101)

In [31]: gg = genlogistic(c=1)

In [32]: kk = kappa4(h=-1, k=0)

In [33]: max(gg.pdf(x) - kk.pdf(x))
Out[33]: 5.5511151231257827e-17
```

Seems to work, no?

timcera replied:
The kappa4 with h=-1 and k=0 is the _logistic_ distribution.  The genlogistic(c=1) distribution is the logistic distribution.

```
In [6]: gg = genlogistic(c = 1)

In [7]: ll = logistic()

In [9]: max(gg.pdf(x) - ll.pdf(x))
Out[9]: 0.0

In [10]: kk = kappa4(h=-1, k=0)

In [11]: max(ll.pdf(x) - kk.pdf(x))
Out[11]: 5.5511151231257827e-17

```

For the genextreme and genpareto distributions I used the _k_ value from the kappa4 in the tests...

For example, the genpareto test:
```
def test_cdf_genpareto(self):
        # h = 1 and k != 0 is generalized Pareto
        x = [0.0, 0.1, 0.2, 0.5]
        h = 1.0
        for k in [-1.9, -1.0, -0.5, -0.2, -0.1, 0.1, 0.2, 0.5, 1.0,
                  1.9]:
            vals = stats.kappa4.cdf(x, h, k)
            # shape parameter is opposite what is expected
            vals_comp = stats.genpareto.cdf(x, -k)
            assert_allclose(vals, vals_comp)
```

I was not able to figure out the relationship between kappa4(h=-1, k=?) and genlogistic(c=?).  In the one case when k=0 in kappa4 and c=1 in genlogistic the distributions are equal and special case to the logistic distribution.  Does this mean that "c" should equal "k + 1"?  Nope, doesn't work.

```
In [12]: kk = kappa4(h=-1, k=1)

In [13]: gg = genlogistic(c = 2)

In [14]: max(gg.pdf(x) - kk.pdf(x))
Out[14]: 0.27343257296753037
```

Basically, according to the literature kappa4(h=-1, k=?) should equal some shaped genlogistic distribution for all k.  

The "fifth" genlogistic equation I mention in the text is the kappa4 equation with h=-1.

ev-br replied:
OK, I see now. This is some other definition of genlogistic. Maybe best to slightly reword the comment though. "Test does not pass" is not the best wording for the docstring. IMO it's better to just state that what's implemented as `genlogistic` in scipy is not what's meant by `genlogistic` in the literature. You have further links to the literature, great.

ev-br commented:
It seems to be almost ready. What's holding it up in my eyes is (1) a slight wording nitpick for the docstring of kappa4, (2) `nan` vs `None` in `kappa4_gen._stats`, and (3) redundant `_lazywhere` calls in kappa3.

The plan is to branch 0.18.x after this weekend. If the changes are in by the branching time, I don't see a problem merging it so that it makes 0.18.0.

timcera commented:
I messed up the rebase.  Will fix.

timcera commented:
Fixed the rebase.

1) Rewrote the documentation to be clearer.
2) I fixed this a while ago, please check.  I had set 'maxr' variable to 0 instead of 5.
3) Fixed.

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -3313,6 +3314,312 @@ def _ppf(self, q, k, s):
+    def _cdf(self, x, a):
+        x = asarray(x)
+        ret = np.zeros(shape(x), dtype=x.dtype)
+        ret[x > 0] = x*(a + x**a)**(-1.0/a)
```
I believe this can still be simplified to just `return x * (a + x**2) ** (-1.0/a)`. Everything else is taken care of automagically by the framework. (possibly with the help of `_open_support_mask`) But this can be cleaned up separately.

ev-br replied:
Actually, no. The CI failure is (you need to click on `get the raw log` at the bottom of the page)
```
======================================================================
ERROR: test_continuous_basic.test_cont_basic(<scipy.stats._continuous_distns.kappa3_gen object at 0x7fcbb8d54278>, (1.0,), 'kappa3')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/scipy/stats/tests/common_tests.py", line 93, in check_entropy
    ent = distfn.entropy(*arg)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/scipy/stats/_distn_infrastructure.py", line 1040, in entropy
    place(output, cond0, self.vecentropy(*goodargs) + log(scale))
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/numpy/lib/function_base.py", line 1874, in __call__
    return self._vectorize_call(func=func, args=vargs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/numpy/lib/function_base.py", line 1940, in _vectorize_call
    outputs = ufunc(*inputs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/scipy/stats/_distn_infrastructure.py", line 2232, in _entropy
    h = integrate.quad(integ, self.a, self.b)[0]
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/scipy/integrate/quadpack.py", line 315, in quad
    points)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/scipy/integrate/quadpack.py", line 382, in _quad
    return _quadpack._qagie(func,bound,infbounds,args,full_output,epsabs,epsrel,limit)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/scipy/stats/_distn_infrastructure.py", line 2227, in integ
    val = self._pdf(x, *args)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/scipy/stats/_continuous_distns.py", line 3605, in _pdf
    ret[x > 0] = a*(a + x**a)**(-1.0/a-1)
IndexError: 0-d arrays can't be indexed.
---------------------------------------------------

ev-br commented:
LGTM now.

ev-br commented:
OK, merged. Thanks @timcera!
